### PR TITLE
Implement macros API

### LINKED ISSUE #797: Implement macros (template variables) API
Add `MacroService` and associated HTTP handler that conforms to the specification

https://github.com/influxdata/platform/blob/f437888ee24e812fb8ddaa03468e7acd1103ef91/http/swagger.yml#L18-L147

https://github.com/influxdata/platform/blob/f437888ee24e812fb8ddaa03468e7acd1103ef91/http/swagger.yml#L2973-L3058

Implementation should be similar to the `ViewService` and `ViewHandler`

### PR DESCRIPTION
**Pull request by chnn**
Closes #797 

Implements the `/v2/macros` API endpoints as described by the [spec](https://github.com/influxdata/platform/blob/8ef01fab293c5f4aee09fd454f570d19bf44b9e9/http/swagger.yml#L18-L147).

### Todo

- [ ] Write up what request bodies should look like in the spec (currently undefined)
- [ ] Add support for the `org` query parameter. I could use some guidance on this (what permissions are associated with an org, how is a user authenticated, what status code for unauthenticated responses, etc.)

### CODE REVIEW

File: macro.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+func (a *MacroArguments) UnmarshalJSON(data []byte) error {
+	type Alias MacroArguments
+	aux := struct{ *Alias }{Alias: (*Alias)(a)}
+
+	err := json.Unmarshal(data, &aux)
+	if err != nil {
+		return err
+	}
+
+	// Decode the polymorphic MacroArguments.Values field into the approriate struct
```
Comment (chnn) [inline macro.go]: I hope there's a better way to do this. Any ideas?

Comment (goller) [reply macro.go]: sadly no.. it's a huge headache.

**Comment by kelwang** [comment]
Hi, @chnn  
Since https://github.com/influxdata/platform/pull/835/files just got merged, could you add inmem for macros as well?

Thanks

**Comment by chnn** [comment]
@kelwang Sure thing.

**Comment by chnn** [comment]
I'm going to follow up with https://github.com/influxdata/platform/issues/837 and  https://github.com/influxdata/platform/issues/838.

**Pull request merged**